crates/provider/src/heart.rs

```diff
@@ -458,37 +504,112 @@ impl<S> Heartbeat<S> {
         let (ix_tx, ixns) = mpsc::channel(16);
 
-        self.into_future(latest, ixns).spawn_task();
+        self.into_future(latest, ixns, client).spawn_task();
 
         HeartbeatHandle { tx: ix_tx, latest: latest_rx }
     }
 }
 
 #[cfg(not(target_arch = "wasm32"))]
-impl<S: Stream<Item = Block> + Unpin + Send + 'static> Heartbeat<S> {
+impl<S: Stream<Item = Block> + Unpin + Send + 'static, N: Network> Heartbeat<S, N> {
     /// Spawn the heartbeat task, returning a [`HeartbeatHandle`].
-    pub(crate) fn spawn(self) -> HeartbeatHandle {
+    pub(crate) fn spawn<T: Transport + Clone>(self, client: WeakClient<T>) -> HeartbeatHandle<N> {
         let (latest, latest_rx) = watch::channel(None::<Block>);
         let (ix_tx, ixns) = mpsc::channel(16);
 
-        self.into_future(latest, ixns).spawn_task();
+        self.into_future(latest, ixns, client).spawn_task();
 
         HeartbeatHandle { tx: ix_tx, latest: latest_rx }
     }
 }
 
-impl<S: Stream<Item = Block> + Unpin + 'static> Heartbeat<S> {
-    async fn into_future(
+type FetchReceiptFuture<R> = Pin<Box<dyn Future<Output = Option<R>> + Send>>;
+
+struct ReceiptTasks<T, N: Network> {
+    client: WeakClient<T>,
+
+    /// Receiver of shutdown signal, used on unexpected channel close.
+    shutdown_tx: mpsc::Sender<()>,
+
+    /// Futures which perform a single request to fetch a receipt.
+    fetch_receipt_tasks: FuturesUnordered<FetchReceiptFuture<N::ReceiptResponse>>,
+
+    /// Futures which poll for receipts of confirmed transactions and notify [TxWatcher]s.
+    poll_receipt_tasks: FuturesUnordered<Pin<Box<dyn Future<Output = ()> + Send>>>,
+}
+
+impl<T: Transport + Clone, N: Network> ReceiptTasks<T, N> {
+    fn new(client: WeakClient<T>, shutdown_tx: mpsc::Sender<()>) -> Self {
+        Self {
+            client,
+            shutdown_tx,
+            fetch_receipt_tasks: Default::default(),
+            poll_receipt_tasks: Default::default(),
+        }
+    }
+
+    fn fetch_receipt(&mut self, tx_hash: B256) {
+        // We don't actually poll, but use poller to retry on failures.
+        let mut poller = PollerBuilder::<_, _, Option<N::ReceiptResponse>>::new(
+            self.client.clone(),
+            "eth_getTransactionReceipt",
+            (tx_hash,),
+        )
+        .spawn()
+        .into_stream();
+
+        let shutdown = self.shutdown_tx.clone();
+
+        self.fetch_receipt_tasks.push(Box::pin(async move {
+            if let Some(receipt) = poller.next().await {
+                receipt
+            } else {
+                let _ = shutdown.send(()).await;
+                None
+            }
+        }));
+    }
+
+    fn poll_receipt(&mut self, watcher: TxWatcher<N>) {
+        let mut poller = PollerBuilder::<_, _, Option<N::ReceiptResponse>>::new(
+            self.client.clone(),
+            "eth_getTransactionReceipt",
+            (watcher.config.tx_hash,),
+        )
+        .spawn()
+        .into_stream();
```

Comment:
hmm do we want to spawn 1 task per tx?

Reply:
those pollers will usually stop after first request because this task is started once tx is included, so I think this is fine?

Reply:
I think the spawn-per-tx falls out of the intent to retry, which probably belongs elsewhere. using the poller to cover up transport failures is an anti-pattern